pcsx2/gui/Debugger/CtrlRegisterList.cpp

```diff
@@ -20,505 +20,436 @@
+            case DebugInterface::NORMAL:
+                cellValue = wxString::Format("%08X", value._u32[numCols - col - 1]);
+                textColor = changed.changed[numCols - col - 1] ? colorChanged : colorUnchanged;
+                break;
+            case DebugInterface::SPECIAL:
+                cellValue = cpu->getRegisterString(cat, row);
+                textColor = (changed.changed[0] || changed.changed[1] || changed.changed[2] || changed.changed[3]) ?
+                    colorChanged : colorUnchanged;
+                break;
+            default: pxAssert(0 && "Unreachable switch case");
+            }
+            if (regGrid->GetCellTextColour(row, col) != textColor)
+                regGrid->SetCellTextColour(row, col, textColor);
+            if (regGrid->GetCellValue(row, col) != cellValue)
+                regGrid->SetCellValue(row, col, cellValue);
+            if (regGrid->GetCellBackgroundColour(row, col) != bgColor)
+                regGrid->SetCellBackgroundColour(row, col, bgColor);
+        }
+    }
+}
+
+void CtrlRegisterList::updateSize(int cat)
+{
+    wxGrid* regGrid = registerGrids[cat];
+
+    // Make sure that the number of shown cols is correct.
+    {
+        int regBits = cpu->getRegisterSize(cat);
+        int numCols = regGrid->GetNumberCols();
+
+        int shownCols = 0;
+        while (shownCols < numCols && regGrid->IsColShown(shownCols)) shownCols++;
+        if (shownCols > maxBits / 32)
+            shownCols = (maxBits / 32);
+        else if (shownCols < regBits / 32)
+            shownCols = std::min(maxBits / 32, regBits / 32);
+
+        for (int col = 0; col < numCols; col++)
+            if (col < shownCols)
+                regGrid->ShowCol(numCols - col - 1); // Big-endian representation so flip order
+            else
+                regGrid->HideCol(numCols - col - 1); // Big-endian representation so flip order
+    }
+
+    regGrid->AutoSize();
+    wxSize pageSize = regGrid->GetSize();
+
+    // Hacks below
+    {
+        // Hack: Sometimes the vertical scroll bar covers some of the text so add some room
+        pageSize.x += 20;
+
+        // Hack: AFAIK wxNotebook does not provide a way to get the necessary size
+        // for the tabs so we use a rough approximation and hope that the tabs
+        // will have enough room to all be shown. 50 pixels per tab should hopefully work.
+        int minX = registerCategories->GetPageCount() * 50;
+        pageSize.x = std::max(pageSize.x, minX);
+
+        // Hack: Sometimes showing all the rows on the screen take up too much
+        // vertical room and squeezes other components (breakpoint window, etc.)
+        // into nothing so we limit the vertical size to 1/2 screen size.
+        // If necessary, this will automatically create a vertical scroll bar so
+        // all rows can be accessed.
+        int screenSize = wxSystemSettings::GetMetric(wxSYS_SCREEN_Y);
+        pageSize.y = std::min(screenSize / 2, pageSize.y);
```

Comment:
Although this is better for small screens, it's provides slightly worse behaviour for larger screens - at the resolution I'm using (1920x1200) I now have to scroll to see all the registers. It probably needs a small adjustment.